src/System.Management.Automation/engine/Subsystem/Prediction/CommandPrediction.cs

```diff
@@ -0,0 +1,170 @@
+        /// <param name="astTokens">The <see cref="Token"/> objects from parsing the current command line input.</param>
+        /// <param name="millisecondsTimeout">The milliseconds to timeout.</param>
+        /// <returns>A list of <see cref="PredictionResult"/> objects.</returns>
+        public static async Task<List<PredictionResult>?> PredictInput(Ast ast, Token[] astTokens, int millisecondsTimeout)
+        {
+            Requires.Condition(millisecondsTimeout > 0, nameof(millisecondsTimeout));
+
+            var predictors = SubsystemManager.GetSubsystems<IPredictor>();
+            if (predictors.Count == 0)
+            {
+                return null;
+            }
+
+            var context = new PredictionContext(ast, astTokens);
+            var tasks = new Task<PredictionResult?>[predictors.Count];
+            using var cancellationSource = new CancellationTokenSource();
+
+            for (int i = 0; i < predictors.Count; i++)
+            {
+                IPredictor predictor = predictors[i];
+
+                tasks[i] = Task.Factory.StartNew(
+                    state => {
+                        var predictor = (IPredictor)state!;
+                        List<PredictiveSuggestion>? texts = predictor.GetSuggestion(context, cancellationSource.Token);
+                        return texts?.Count > 0 ? new PredictionResult(predictor.Id, predictor.Name, texts) : null;
+                    },
+                    predictor,
+                    CancellationToken.None,
+                    TaskCreationOptions.DenyChildAttach,
+                    TaskScheduler.Default);
+            }
+
+            await Task.WhenAny(
+                Task.WhenAll(tasks),
+                Task.Delay(millisecondsTimeout, cancellationSource.Token)).ConfigureAwait(false);
+            cancellationSource.Cancel();
+
+            var results = new List<PredictionResult>(predictors.Count);
+            foreach (Task<PredictionResult?> task in tasks)
+            {
+                if (task.IsCompletedSuccessfully)
+                {
+                    PredictionResult? result = task.Result;
+                    if (result != null)
+                    {
+                        results.Add(result);
+                    }
+                }
+            }
+
+            return results;
+        }
+
+        /// <summary>
+        /// Allow registered predictors to do early processing when a command line is accepted.
+        /// </summary>
+        /// <param name="history">History command lines provided as references for prediction.</param>
+        public static void OnCommandLineAccepted(IReadOnlyList<string> history)
+        {
+            Requires.NotNull(history, nameof(history));
+
+            var predictors = SubsystemManager.GetSubsystems<IPredictor>();
+            if (predictors.Count == 0)
+            {
+                return;
+            }
+
+            foreach (IPredictor predictor in predictors)
```

Comment:
Why not? Otherwise, an enumerator of `predictors` will be created every time this method is called, even though `predictors` is empty. Yeah, I know you can change that to `for` loop, but I like the current code better.